plugin/rewrite/reverter.go

```diff
@@ -21,23 +22,69 @@ type ResponseRule struct {
 // dig will complain with ';; Question section mismatch: got example.org/HINFO/IN'
 type ResponseReverter struct {
 	dns.ResponseWriter
-	originalQuestion dns.Question
-	ResponseRewrite  bool
-	ResponseRules    []ResponseRule
+	request           *dns.Msg
+	originalQuestion  dns.Question
+	originalOPTHeader dns.RR_Header
+	ResponseRewrite   bool
+	ResponseRules     []ResponseRule
+}
+
+// align the OPT inside the msg the information provided - hasOPT and Header
+// the sub array of EDNS0 is not changed (unless OPT is removed)
+func replaceOPTHeader(m *dns.Msg, hasOpt bool, hdr dns.RR_Header) {
+	for i := len(m.Extra) - 1; i >= 0; i-- {
+		if m.Extra[i].Header().Rrtype == dns.TypeOPT {
+			if !hasOpt {
+				copy(m.Extra[i:], m.Extra[i+1:])
+				m.Extra[len(m.Extra)-1] = nil
+				m.Extra = m.Extra[:len(m.Extra)-1]
+				return
+			}
+			m.Extra[i].(*dns.OPT).Hdr = hdr
+			return
+		}
+	}
+	if hasOpt {
+		opt := dns.OPT{hdr, make([]dns.EDNS0, 0)}
+		m.Extra = append(m.Extra, &opt)
+	}
 }
 
 // NewResponseReverter returns a pointer to a new ResponseReverter.
 func NewResponseReverter(w dns.ResponseWriter, r *dns.Msg) *ResponseReverter {
-	return &ResponseReverter{
+	rr := &ResponseReverter{
 		ResponseWriter:   w,
+		request:          r,
 		originalQuestion: r.Question[0],
 	}
+	if o := r.IsEdns0(); o != nil {
+		rr.originalOPTHeader = *o.Header()
+	}
+	return rr
+}
+
+func (r *ResponseReverter) hasOriginalOPT() bool {
+	return r.originalOPTHeader.Rrtype == dns.TypeOPT
 }
 
 // WriteMsg records the status code and calls the underlying ResponseWriter's WriteMsg method.
 func (r *ResponseReverter) WriteMsg(res *dns.Msg) error {
+
+	// set back the client request to its original OPT header
+	replaceOPTHeader(r.request, r.hasOriginalOPT(), r.originalOPTHeader)
+	if !r.hasOriginalOPT() {
+		// ensure remove OPT out of the response
+		// if initial query was not EDNS0 - compliance with https://tools.ietf.org/html/rfc6891#page-10
+		replaceOPTHeader(res, r.hasOriginalOPT(), r.originalOPTHeader)
+	}
+
+	// always revert the question
 	res.Question[0] = r.originalQuestion
+
+	// Revert part happen ONLY if flag is set
```

Comment:
This comment doesn't add anything (also misses closing dot)

Reply:
i'll remove then.